File: src/test/java/io/pivotal/kotlindemo/java/accounts/FakeAccountRepository.java
package io.pivotal.kotlindemo.java.accounts;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class FakeAccountRepository implements AccountRepository {

    private Map<Long, Account> accountMap = new HashMap<>();

    @Override
    public Account create(CreateAccountRequest createAccountRequest) {
        Long id = (long) accountMap.values().size();
        Account account = new Account(
                id,
                createAccountRequest.getClientId(),
                createAccountRequest.getNumber(),
                createAccountRequest.getFullname(),
                createAccountRequest.getNickname(),
                createAccountRequest.getType());

        accountMap.put(id, account);

        return account;
    }

    @Override
    public List<Account> findAll() {
        return new ArrayList<>(accountMap.values());
    }

    @Override
    public Account findById(Long id) {
        Account account = accountMap.get(id);
        if (account == null) {
            throw new NoAccountFoundException();
        }
        return account;
    }

    @Override
    public List<Account> findByClientId(Long id) {
        return accountMap
                .values()
                .stream()
                .filter(account -> account.getClientId().equals(id))
                .collect(Collectors.toList());
    }

}

File: src/test/kotlin/io/pivotal/kotlindemo/kotlin/accounts/FakeAccountRepository.kt
package io.pivotal.kotlindemo.kotlin.accounts

class FakeAccountRepository: AccountRepository {
    private val accountsMap = HashMap<Long, Account>()

    override fun create(createAccountRequest: CreateAccountRequest): Account {
        val id = accountsMap.size.toLong()
        val account = Account(
            id = id,
            clientId = createAccountRequest.clientId,
            number = createAccountRequest.number,
            fullname = createAccountRequest.fullname,
            nickname = createAccountRequest.nickname,
            type = createAccountRequest.type
        )
        accountsMap.put(id, account)
        return account
    }

    override fun findAll(): List<Account> {
        return accountsMap.values.toList()
    }

    override fun findById(id: Long): Account {
        return accountsMap[id] ?: throw NoAccountFoundException()
    }

    override fun findByClientId(clientId: Long): List<Account> {
        return accountsMap.values.filter { it -> it.clientId == clientId }
    }

}
